external/java/src/main/java/ch/epfl/dedis/lib/crypto/bn256/BN.java

```diff
@@ -0,0 +1,521 @@
+package ch.epfl.dedis.lib.crypto.bn256;
+
+import java.math.BigInteger;
+import java.util.Arrays;
+import java.util.Random;
+
+public class BN {
+    /**
+     * The pair of a G1 point and a scalar that is used to generate it.
+     */
+    public static class PairG1 {
+        public BigInteger k;
+        public G1 p;
```

Comment:
I know you want to keep stuff as the go implementation but it might be worth to make that private if it doesn't require to many changes (honestly I didn't check). It's just that you can't really change the attributes as they're bounded together. I leave a comment only here so you decide if it's useful or not for here and other places.